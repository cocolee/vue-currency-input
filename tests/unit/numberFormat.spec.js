import NumberFormat from '../../src/numberFormat'

describe('NumberFormat', () => {
  describe('constructing number formats', () => {
    it('should work for respective locale and currency', () => {
      expect(new NumberFormat({ locale: 'de-DE', currency: 'EUR' })).toMatchSnapshot('de-DE_EUR')
      expect(new NumberFormat({ locale: 'de-CH', currency: 'EUR' })).toMatchSnapshot('de-CH_EUR')
      expect(new NumberFormat({ locale: 'es-ES', currency: 'EUR' })).toMatchSnapshot('es-ES_EUR')
      expect(new NumberFormat({ locale: 'nl-NL', currency: 'EUR' })).toMatchSnapshot('nl-NL_EUR')
      expect(new NumberFormat({ locale: 'en-US', currency: 'USD' })).toMatchSnapshot('en-US_USD')
      expect(new NumberFormat({ locale: 'fr-CH', currency: 'CHF' })).toMatchSnapshot('fr-CH_CHF')
      expect(new NumberFormat({ locale: 'zh', currency: 'CNY' })).toMatchSnapshot('zh_CNY')
      expect(new NumberFormat({ locale: 'en-GB', currency: 'GBP' })).toMatchSnapshot('en-GB_GBP')
      expect(new NumberFormat({ locale: 'en-GB', currency: 'INR' })).toMatchSnapshot('en-IN_INR')
      expect(new NumberFormat({ locale: 'pt', currency: 'BRL' })).toMatchSnapshot('pt_BRL')
      expect(new NumberFormat({ locale: 'ja', currency: 'JPY' })).toMatchSnapshot('ja_JPY')
      expect(new NumberFormat({ locale: 'ar-SA', currency: 'SAR' })).toMatchSnapshot('ar-SA_SAR')
      expect(new NumberFormat({ locale: 'fa-IR', currency: 'IRR' })).toMatchSnapshot('fa-IR_IRR')
    })

    describe('custom currency symbol', () => {
      it('should work with a custom prefix', () => {
        expect(new NumberFormat({ locale: 'de-DE', currency: { prefix: '₿ ' } })).toMatchSnapshot()
      })

      it('should work wit a custom suffix', () => {
        expect(new NumberFormat({ locale: 'de-DE', currency: { suffix: ' Euro' } })).toMatchSnapshot()
      })

      it('should work if the currency symbol is hidden', () => {
        expect(new NumberFormat({ locale: 'de-DE' })).toMatchSnapshot()
      })
    })

    describe('custom precision', () => {
      it('should work with a custom precision', () => {
        expect(new NumberFormat({ precision: 0 })).toEqual(expect.objectContaining({ minimumFractionDigits: 0, maximumFractionDigits: 0 }))
        expect(new NumberFormat({ precision: { min: 0, max: 0 } })).toEqual(expect.objectContaining({ minimumFractionDigits: 0, maximumFractionDigits: 0 }))
        expect(new NumberFormat({ precision: { min: 0, max: 2 } })).toEqual(expect.objectContaining({ minimumFractionDigits: 0, maximumFractionDigits: 2 }))
      })

      it('should ignore precision ranges when "Auto decimal mode" is used', () => {
        expect(new NumberFormat({ currency: 'EUR', autoDecimalMode: true, precision: { min: 0, max: 2 } })).toEqual(expect.objectContaining({ minimumFractionDigits: 2, maximumFractionDigits: 2 }))
      })

      it('should ignore precision ranges when "Value as integer" is used', () => {
        expect(new NumberFormat({ currency: 'EUR', valueAsInteger: true, precision: { min: 0, max: 2 } })).toEqual(expect.objectContaining({ minimumFractionDigits: 2, maximumFractionDigits: 2 }))
      })

      it('should ignore the custom precision if the locale does not support decimal digits', () => {
        expect(new NumberFormat({ locale: 'ja', currency: 'JPY', precision: 2 })).toEqual(expect.objectContaining({ minimumFractionDigits: 0, maximumFractionDigits: 0 }))
      })
    })
  })

  describe('parse', () => {
    it('should return null if the value is empty', () => {
      expect(new NumberFormat({ locale: 'en' }).parse(null)).toBeNull()
      expect(new NumberFormat({ locale: 'en' }).parse('')).toBeNull()
      expect(new NumberFormat({ locale: 'en' }).parse(' ')).toBeNull()
    })

    it('should return null if the value is invalid', () => {
      expect(new NumberFormat({ locale: 'en' }).parse('-')).toBeNull()
      expect(new NumberFormat({ locale: 'en' }).parse('123e-1')).toBeNull()
      expect(new NumberFormat({ locale: 'en' }).parse('0x11')).toBeNull()
      expect(new NumberFormat({ locale: 'en' }).parse('0b11')).toBeNull()
      expect(new NumberFormat({ locale: 'en' }).parse('0o11')).toBeNull()
      expect(new NumberFormat({ locale: 'en' }).parse('1.2e1')).toBeNull()
      expect(new NumberFormat({ locale: 'en' }).parse('1.23.4')).toBeNull()
    })

    it('should return the parsed number if the value conforms to the currency format', () => {
      expect(new NumberFormat({ locale: 'en' }).parse('1234')).toBe(1234)
      expect(new NumberFormat({ locale: 'en' }).parse('1,234,567')).toBe(1234567)
      expect(new NumberFormat({ locale: 'en' }).parse('-1,234,567')).toBe(-1234567)
      expect(new NumberFormat({ locale: 'de' }).parse('-1234567,89')).toBe(-1234567.89)
      expect(new NumberFormat({ locale: 'en', currency: 'USD' }).parse('$1,234,567')).toBe(1234567)
      expect(new NumberFormat({ locale: 'de', currency: 'EUR' }).parse('1234 €')).toBe(1234)
      expect(new NumberFormat({ locale: 'en' }).parse('-1234')).toBe(-1234)
      expect(new NumberFormat({ locale: 'en', currency: 'USD' }).parse('-$1234')).toBe(-1234)
      expect(new NumberFormat({ locale: 'de', currency: 'EUR' }).parse('-1234 €')).toBe(-1234)
      expect(new NumberFormat({ locale: 'ja', currency: 'JPY' }).parse('￥123,456')).toBe(123456)
      expect(new NumberFormat({ locale: 'en' }).parse('0.5')).toBe(0.5)
      expect(new NumberFormat({ locale: 'en' }).parse('1234.50')).toBe(1234.5)
      expect(new NumberFormat({ locale: 'en' }).parse('1234.00')).toBe(1234)
      expect(new NumberFormat({ locale: 'en', currency: 'USD' }).parse('$1,234.50')).toBe(1234.5)
      expect(new NumberFormat({ locale: 'de', currency: 'EUR' }).parse('1.234,50 €')).toBe(1234.5)
      expect(new NumberFormat({ locale: 'ar', currency: 'SAR' }).parse('١٢٣٤')).toBe(1234)
      expect(new NumberFormat({ locale: 'ar', currency: 'SAR' }).parse('١٬٢٣٤')).toBe(1234)
      expect(new NumberFormat({ locale: 'ar', currency: 'SAR' }).parse('٠٫٩')).toBe(0.9)
      expect(new NumberFormat({ locale: 'ar', currency: 'SAR' }).parse('؜-٠٫٥٠ ر.س.‏')).toBe(-0.5)
      expect(new NumberFormat({ locale: 'en-IN', currency: 'INR' }).parse('₹1,23,334.00')).toBe(123334)
      expect(new NumberFormat({ locale: 'de-AT', currency: 'EUR' }).parse('€ 66.668')).toBe(66668)
      expect(new NumberFormat({ locale: 'de-AT', currency: null }).parse('-66 668')).toBe(-66668)
    })

    it('should return null if the value does not conform to the currency format', () => {
      expect(new NumberFormat({ locale: 'en' }).parse('1234,5')).toBeNull()
      expect(new NumberFormat({ locale: 'de' }).parse('1,234,567.89')).toBeNull()
      expect(new NumberFormat({ locale: 'de', currency: 'EUR' }).parse('$1234')).toBeNull()
      expect(new NumberFormat({ locale: 'en', currency: 'USD' }).parse('1234 €')).toBeNull()
      expect(new NumberFormat({ locale: 'ja', currency: 'JPY' }).parse('1234.56')).toBeNull()
    })
  })

  describe('isFractionIncomplete', () => {
    it('should return true if the fraction is incomplete', () => {
      expect(new NumberFormat({ locale: 'en' }).isFractionIncomplete('1234.')).toBe(true)
      expect(new NumberFormat({ locale: 'en' }).isFractionIncomplete('1234')).toBe(false)
      expect(new NumberFormat({ locale: 'en' }).isFractionIncomplete('1234.3')).toBe(false)
      expect(new NumberFormat({ locale: 'en' }).isFractionIncomplete('1234.3.')).toBe(false)
      expect(new NumberFormat({ locale: 'de' }).isFractionIncomplete('1,3,')).toBe(false)
    })
  })

  describe('normalizeDecimalSymbol', () => {
    it('should replace the first decimal symbol with the one of the current locale', () => {
      expect(new NumberFormat({ locale: 'en' }).normalizeDecimalSymbol('1,23,4,567', 2)).toBe('1,23.4,567')
    })
  })

  describe('format', () => {
    it('should return the formatted value for the respective options', () => {
      expect(new NumberFormat({ locale: 'en', currency: 'EUR' }).format(1234.5789)).toBe('€1,234.58')
      expect(new NumberFormat({ locale: 'en', currency: 'EUR' }).format(1234.5789, { minimumFractionDigits: 4 })).toBe('€1,234.5789')
      expect(new NumberFormat({ locale: 'en' }).format(-1)).toBe('-1.00')
      expect(new NumberFormat({ locale: 'en' }).format(-0)).toBe('-0.00')
      expect(new NumberFormat({ locale: 'en' }).format(1234.5789, { minimumFractionDigits: 4 })).toBe('1,234.5789')
    })
  })
})
