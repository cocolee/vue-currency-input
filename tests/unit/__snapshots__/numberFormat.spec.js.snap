// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberFormat constructing number formats custom currency symbol should work if the currency symbol is hidden 1`] = `
NumberFormat {
  "currency": undefined,
  "decimalSymbol": ",",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "de-DE",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats custom currency symbol should work wit a custom suffix 1`] = `
NumberFormat {
  "currency": Object {
    "suffix": " Euro",
  },
  "decimalSymbol": ",",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "de-DE",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " Euro",
}
`;

exports[`NumberFormat constructing number formats custom currency symbol should work with a custom prefix 1`] = `
NumberFormat {
  "currency": Object {
    "prefix": "₿ ",
  },
  "decimalSymbol": ",",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "de-DE",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-₿ ",
  "prefix": "₿ ",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: ar-SA_SAR 1`] = `
NumberFormat {
  "currency": "SAR",
  "decimalSymbol": "٫",
  "digits": Array [
    "٠",
    "١",
    "٢",
    "٣",
    "٤",
    "٥",
    "٦",
    "٧",
    "٨",
    "٩",
  ],
  "groupingSymbol": "٬",
  "locale": "ar-SA",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "؜-",
  "negativePrefix": "؜-",
  "prefix": "",
  "suffix": " ر.س.‏",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: de-CH_EUR 1`] = `
NumberFormat {
  "currency": "EUR",
  "decimalSymbol": ".",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": "’",
  "locale": "de-CH",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "EUR-",
  "prefix": "EUR ",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: de-DE_EUR 1`] = `
NumberFormat {
  "currency": "EUR",
  "decimalSymbol": ",",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "de-DE",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: en-GB_GBP 1`] = `
NumberFormat {
  "currency": "GBP",
  "decimalSymbol": ".",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "en-GB",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-£",
  "prefix": "£",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: en-IN_INR 1`] = `
NumberFormat {
  "currency": "INR",
  "decimalSymbol": ".",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "en-GB",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-₹",
  "prefix": "₹",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: en-US_USD 1`] = `
NumberFormat {
  "currency": "USD",
  "decimalSymbol": ".",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "en-US",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-$",
  "prefix": "$",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: es-ES_EUR 1`] = `
NumberFormat {
  "currency": "EUR",
  "decimalSymbol": ",",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "es-ES",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: fa-IR_IRR 1`] = `
NumberFormat {
  "currency": "IRR",
  "decimalSymbol": undefined,
  "digits": Array [
    "۰",
    "۱",
    "۲",
    "۳",
    "۴",
    "۵",
    "۶",
    "۷",
    "۸",
    "۹",
  ],
  "groupingSymbol": "٬",
  "locale": "fa-IR",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minusSymbol": "‎−",
  "negativePrefix": "‎−‎ریال ",
  "prefix": "‎ریال ",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: fr-CH_CHF 1`] = `
NumberFormat {
  "currency": "CHF",
  "decimalSymbol": ".",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": " ",
  "locale": "fr-CH",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " CHF",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: ja_JPY 1`] = `
NumberFormat {
  "currency": "JPY",
  "decimalSymbol": undefined,
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "ja",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "minusSymbol": "-",
  "negativePrefix": "-￥",
  "prefix": "￥",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: nl-NL_EUR 1`] = `
NumberFormat {
  "currency": "EUR",
  "decimalSymbol": ",",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "nl-NL",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "€ -",
  "prefix": "€ ",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: pt_BRL 1`] = `
NumberFormat {
  "currency": "BRL",
  "decimalSymbol": ",",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ".",
  "locale": "pt",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-R$ ",
  "prefix": "R$ ",
  "suffix": "",
}
`;

exports[`NumberFormat constructing number formats should work for respective locale and currency: zh_CNY 1`] = `
NumberFormat {
  "currency": "CNY",
  "decimalSymbol": ".",
  "digits": Array [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9",
  ],
  "groupingSymbol": ",",
  "locale": "zh",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "minusSymbol": "-",
  "negativePrefix": "-¥",
  "prefix": "¥",
  "suffix": "",
}
`;
